RFC-0004-sparse-fill-value/README.md

```diff
@@ -0,0 +1,613 @@
+    B = [[*, 6], [*, 8]]                  # B fill value is 7
+    A + B = [[1 + 3, 2 + 4], [*, 6 + 8]]  # A + B fill value is 2 + 7 = 9
+    ```
+
+10. Existing PyTorch functions that support sparse tensor as inputs,
+    need to be updated for handling the defined fill values. This will
+    be implemented in two stages:
+
+    1. All relevant functions need to check for zero fill value. If a
+       nonzero fill value is used, raise an exception.
+
+       For instance:
+
+       ```C++
+       TORCH_CHECK(A._fill_value().nonzero().numel() == 0,
+                   "The <function> requires a sparse tensor with zero fill value, got ",
+                   A.fill_value());
+       ```
+
+       This check ensures that the existing functions will not produce
+       incorrect results silently when users pass a sparse tensor
+       with nonzero fill value as an input while the function assumes
+       it to be zero.
+
+    2. Update the related functions to handle nonzero fill values of
+       input sparse tensors correctly.
+
+       For instance, consider a matrix multiplication of two sparse
+       tensors `A` and `B` with fill values `a` and `b`, respectively,
+       then the `matmul` operation can be expanded as follows:
+
+       ```python
+       matmul(A, B) = matmul(A - fA + fA, B - fB + fB)
+                    = matmul(A - fA, B - fB) + fA * matmul(ones_like(A), B) + fB * matmul(A, ones_like(B))
+       ```
+
+       where the first term can be computed using existing matmul for
+       sparse tensors with zero fill value, and the last two terms can
+       be replaced with a computation of a single row or column of the
+       corresponding matrix products that has reduced computational
+       complexity.
+
+11. We propose to add an optional argument `fill_value` to the `to_sparse`
+    method:
+
+    ```python
+    torch.Tensor.to_sparse(self, sparseDims=None, fill_value=None)
+    ```
+
+    that enables efficient construction of sparse tensors from tensors
+    with repeated values equal to given `fill_value`. For example:
+
+    ```python
+    torch.ones(10).to_sparse(fill_value=1.) -> torch.sparse_coo_tensor([[]], [], (10,), fill_value=1.)
+    ```
+
+    The `fill_value` argument has the same semantics as in sparse
+    tensor constructors (see point 1) and its shape must be consistent
+    with the shape of the input tensor (`self`) and the value of
+    `sparseDims` when specified: `self.shape[:sparseDims]` must be
+    equal to `<constructed sparse tensor>.fill_value().shape`.
+
+12. Sparse tensors with defined fill value have intrinsic constraints
+    between all the unspecified tensor elements (these are always
+    equal) that must be taken into account when implementing Autograd
+    backward methods for functions that receive sparse tensors as
+    inputs.
+
+    Sparse tensors with indefinite fill value don't have the intrinsic
+    constraints as discussed above.
```

Comment:
Does autograd even make sense in that case? I suspect not, in which case just deleting this sentence would be good.

Reply:
I think it does if Graph models would need the autograd support. But it makes to delete this as the proposal does not tackle all the details that will be related to indefinite fill value.

Reply:
deleted